source/common/buffer/watermark_buffer.h

```diff
@@ -36,6 +36,7 @@ class WatermarkBuffer : public OwnedImpl {
   void move(Instance& rhs, uint64_t length) override;
   SliceDataPtr extractMutableFrontSlice() override;
   uint64_t reserve(uint64_t length, RawSlice* iovecs, uint64_t num_iovecs) override;
+  Reservation reserve(uint64_t preferred_length) override;
```

Comment:
Yes, this PR still needs API comments, and some new tests. I'm just trying to push stuff out as I write it in case reviewers dislike the direction it's going.